A240 refactor mapping tool

emrojo opened the pull request.

emrojo committed:
React JSX transformation for babel
React support for rails + webpack
ExcelToState transformer to generate what it will be the state of the
React component ManifestEditor
Provenance view loads the ManifestEditor component
Webpacker config to process JSX
Updated package.json to load Babel, React and  Redux
Upload controller will generate a valid state for React mapping tool

emrojo committed:
Starting test for incomplete manifest and mapping

emrojo committed:
Added file to test mapping tool display

emrojo committed:
FactoryBot update - Static attributes are not supported anymore, so code
is updated to dynamic attributes

emrojo committed:
Load mapping decisions from server
Lets the user decides the rest of options
Updates table with the mapping provided

emrojo committed:
Schema not rightly duplicated

emrojo committed:
Excel to state will get the manifest model object to be able to
generate a valid state
Renaming of some fields in the manifest state for React
Manifest_id will be sent along with the manifest file
Manifest::Schema module to be able to generate a customized schema for
the specific manifest we are creating. This will let us know define
attributes differently depending on the way the manifest has been declared
(if it is a tube, or is a plate it will ask for some attributes in a
different way). The configuration of this schema generation will be provided
in a .yml file declaring:
 - Field to identify different labwares in the manifest (plate_id)
 - Field to identify different positions in the labware (position)
 - Description for new schema properties not declared for the material
 service, as this manifest schema is going to have slightly different content
from the materials schema

A column will be tried to be matched if:
- It is meant to be shown and has a regular expression associated.
    --- or ---
- It is required

If it is not matched, it will appear in the Expected column of the mapping
tool and the mapping editor will be opened so the user can take actions about
it.

emrojo committed:
Manifest schema generation

emrojo committed:
Top level reducer to be able to perform a validation using the status of the
object in lower branches of the state
Selected options moved to the state for the mapping tool

emrojo committed:
Passes all tests

emrojo committed:
Added waiting times

emrojo committed:
Mapping service to generate mapping and perform update of the labware using
the provenance service.

emrojo committed:
Provenance state split
Removed mapping service

emrojo committed:
Removed service

emrojo committed:
Added attribugte shown and valid

emrojo committed:
Load manifest action

emrojo committed:
Move some error messages to the server

emrojo committed:
Added more error handling

emrojo committed:
Tests

emrojo committed:
Generates the tab bar

emrojo committed:
Displaying values in manifest

emrojo committed:
Changes in the property passing to avoid too many rerenders

emrojo committed:
Start replacing old table. Taxonomy id control seems to be compatible so
I'll leave it for the last change

emrojo committed:
Added redux-saga to handle async actions in redux
Select will match options ignoring case on rendering

emrojo committed:
Save and restore tab

emrojo committed:
Added classnames to manage better setting of specific classes for html

emrojo committed:
Added selectors

emrojo committed:
Optimizing response time for ui

emrojo committed:
Tests updated

emrojo committed:
Remove unneeded files

emrojo committed:
Performance debugging

emrojo committed:
Debounced onChange event that saves the full manifest when a timeout of 500ms
withouth changes in the current input happens

emrojo committed:
Corrected default initial values and behaviour on empty updates

emrojo committed:
Input components to change between using Redux or React state to be able
to accept modification of its content from other parts of the app

emrojo committed:
Taxonomy id control inside react

emrojo committed:
Merged rest of components.

emrojo committed:
Fix saving of materials

emrojo committed:
Passing js tests

emrojo committed:
Corrected naming for set_biomaterial_data tests

emrojo committed:
Some more tests adapted

emrojo committed:
Fixed tests

emrojo committed:
More code removal

emrojo committed:
Removing unneeded files

emrojo committed:
Some more files removal

emrojo committed:
Upload manifest from jquery component

emrojo committed:
Moved all react code under its own folder

emrojo committed:
Change in test running line

pjvv commented on app/javascript/react/actions.es6:
```diff
@@ -0,0 +1,191 @@
+        return
+      }
+    }
+    return $.ajax(taxonomyServiceUrl+'/'+taxId, {
```
Found where you use the service URL!

emrojo replied:
Thanks, for some reason I thought this was different. Then it is still needed

pjvv commented on app/javascript/react/components/labware_tabs.jsx:
```diff
@@ -0,0 +1,80 @@
+import { changeTab, saveTab } from '../actions'
+import classNames from 'classnames'
+
+const logName = (name) => { }
```
Is logging only enabled for debug mode?

emrojo replied:
I was using it for debugging. It should not be there. I'll remove it.

pjvv replied:
I think it's good to keep it for debugging and developing, maybe including a flag that will only show it when not in production?

pjvv commented on app/javascript/react/components/manifest_editor.jsx:
```diff
@@ -0,0 +1,247 @@
+    store.dispatch(loadManifest(props))
+  }
+
+  $(document.body).on('uploadManifest', (event, request) => {
```
Could we swap jQuery out for something else?

emrojo replied:
This piece of code is not needed anymore. uploadManifest() is called from the store, not from a trigger as it was before. I'll remove it.

pjvv replied:
Great, less jQuery! :D

pjvv commented on app/javascript/react/components/mapping_tool.jsx:
```diff
@@ -0,0 +1,251 @@
+      </div>
+      <div className="col-md-5">
+        <div className="form-group">
+          <label htmlFor="fields-from-csv">Fields from CSV</label>
```
Since we are allowing both CSV and Excel files, we should probably update the instructions and references in this modal dialog.

emrojo replied:
Ok

pjvv commented on app/javascript/react/components/mapping_tool.jsx:
```diff
@@ -0,0 +1,251 @@
+    this.state = {
+    }
+  }
+  componentDidUpdate(){
```
This method looks unused.

emrojo replied:
Yes, this component can now be just a function. I'll change it

pjvv commented on app/javascript/react/constants.es6:
```diff
@@ -0,0 +1,20 @@
+const C = {
```
Why are some files saved as .es6 and others as .js?

emrojo replied:
.es6 for javascript using ecmascript 6
.jsx for any javascript that uses React jsx in it
.js for any vanilla javascript

pjvv replied:
Unless webpack or a transpiler handles .js and .es6 files differently, won't it be less confusing to simply call them both .js files?

emrojo replied:
I thought it gives a bit more of information for knowing which type of syntax you have to follow in the file when you open it to do some changes, but I don't have any strong opinion about it. If you think it is confusing to have this type of differentiation I'll change it.

pjvv replied:
I see. I think it's safe to assume that we should use es6 syntax throughout and if we need to add babel at a later stage it would still be worth it?

emrojo replied:
Ok

pjvv commented on app/javascript/react/components/mapping_tool.jsx:
```diff
@@ -0,0 +1,251 @@
+        data-dismiss="modal" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+      </button>
+      <Modal.Title>Select CSV mappings</Modal.Title>
```
Since we are allowing CSV and Excel files, we should probably update instructions and references.

pjvv commented on app/javascript/react/reducers/content.js:
```diff
@@ -0,0 +1,112 @@
+  return newState
+}
+
+// const resetPreviousValues = (prevState, newState) => {
```
This comment can be safely removed.

pjvv commented on app/javascript/react/reducers/mapping.es6:
```diff
@@ -0,0 +1,34 @@
+    case C.MATCH_SELECTION:
+      let mp = Object.assign({}, state)
+      mp.matched.push({observed: action.observed, expected: action.expected})
+      mp.observed=mp.observed.filter((elem) => {return elem != action.observed})
```
Should we be using !== instead of != to prevent type coercion here?

pjvv commented on app/javascript/react/reducers/mapping.es6:
```diff
@@ -0,0 +1,34 @@
+    case C.UNMATCH:
+      let mapping = Object.assign({}, state)
+      mapping.matched = mapping.matched.filter((elem) => {
+        return !((elem.observed == action.observed) && (elem.expected == action.expected))
```
Similarly, === instead of ==?

pjvv commented on app/models/manifest/provenance_state/content_accessor.rb:
```diff
@@ -0,0 +1,164 @@
+require 'active_support/core_ext/module/delegation'
```
Why is this required here and not in the other places delegate is used?

pjvv commented on config/webpack/environment.js:
```diff
@@ -2,6 +2,8 @@ const { environment } = require('@rails/webpacker')
 const erb =  require('./loaders/erb')
 const webpack = require('webpack')
 
+//const Reception = require('./routes.js.erb')
```
Can remove this.

emrojo replied:
Ok

pjvv commented on config/webpack/environment.js:
```diff
@@ -2,6 +2,8 @@ const { environment } = require('@rails/webpacker')
 
+//const Reception = require('./routes.js.erb')
+
 // [PJ] add jQuery and papaparse here so that it remains available to all JS files
```
It seems we do not require papaparse anymore?

emrojo replied:
Ok

pjvv commented on spec/factories/labware_types.rb:
```diff
@@ -1,43 +1,43 @@
-    num_of_rows 8
-    col_is_alpha false
-    row_is_alpha false
+    num_of_cols { 12 }
```
Why are we defining blocks for these variables?

emrojo replied:
I had some problems while running the tests that for some reason were solved after changing this to blocks

pjvv replied:
Haha, weird!

emrojo replied:
It may be possible that after updating some libraries it was in a version of factorybot that deprecated this (possibly as they say here  https://robots.thoughtbot.com/deprecating-static-attributes-in-factory_bot-4-11).  
It was in commit: 4f264c8bef251a5ee5930ebb8d6e645f1f58786a

pjvv replied:
👍

pjvv commented on spec/features/taxon_id_finder_spec.rb:
```diff
@@ -80,31 +80,32 @@
+            fill_in('labware[0]address[A:1]fieldName[taxon_id]', with: '3')
+            expect(page).to have_selector("input[value='Some specie name']")
+
+            #expect(page.find('labware[0]address[A:1]fieldName[scientific_name]')).not_to have_selector('.has-error')
```
Can we remove this test?

emrojo replied:
Do you mean the line or the whole test?

pjvv replied:
Just the line..

pjvv commented on app/javascript/react/actions.es6:
```diff
@@ -0,0 +1,191 @@
+        return
+      }
+    }
+    return $.ajax(taxonomyServiceUrl+'/'+taxId, {
```
Is the taxon ID being validated on the client and in the server?

emrojo replied:
Yes, it seems.

emrojo committed:
Some changes from code review

emrojo committed:
Change of extension: .es6 to .js

emrojo committed:
Added eslint to codeclimate

emrojo committed:
Eslint applied fixes

emrojo committed:
Optional chaining support for eslint with babel

emrojo committed:
eslint changes

emrojo committed:
Finish eslint - default rules

pjvv commented:
If we are using yarn we should and safely can remove the `package-lock.json` file. Would you agree?

emrojo commented:
Ok, removing package-lock.json

emrojo committed:
Removed package-lock.json

emrojo committed:
Codebase a240 refactor mapping tool (#345)

* Added eslint to codeclimate

* Eslint applied fixes

* Optional chaining support for eslint with babel

* eslint changes

* Finish eslint - default rules

* Removed package-lock.json

pjvv commented on app/javascript/react/components/labware_content_input.jsx:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  setStateValueSelected(str) {
+    console.log('set value to '+str)
```
The logging in this file should be wrapped in a debug/devel environment checker or removed.

pjvv committed:
Run bundle update

pjvv commented:
When trying to cancel a manifest, the error `route matches [DELETE] "/manifests/1"` is shown. Should the delete URL also be sent along similarly to the save manifest?

pjvv commented:
When uploading a file, the URL also needs to be corrected.

emrojo committed:
Some codeclimate changes

emrojo committed:
Merge branch 'a240_refactor_mapping_tool' of https://github.com/sanger/aker-reception-app into CODEBASE_a240_refactor_mapping_tool

emrojo committed:
Renaming / Testing execution changes

emrojo committed:
Refactoring cognitive complexity

emrojo committed:
Bugfixing codeclimate refactors

emrojo committed:
Some more refactors

emrojo committed:
Merge pull request #347 from sanger/CODEBASE_a240_refactor_mapping_tool

Codebase a240 refactor mapping tool

emrojo committed:
Added scrollbar to be able to use better the mapping tool
Bugfix: manifestId was not being sent because of removing a previous
content from the views

emrojo committed:
Changed the way rebuilding mapping and content works by using a rebuild flag
instead of checking other different values inside the manifest state object.

emrojo committed:
Added babel syntax plugin to normal dependencies

emrojo committed:
Moved some libraries from devDependencies into dependencies

emrojo committed:
pry moved to the development section in Gemfil

emrojo committed:
Removed pry

emrojo committed:
Added pry

emrojo committed:
Removed pry and pry-rails

emrojo committed:
Readded pry

emrojo committed:
Set path for update state using js routs

emrojo committed:
First change to reload state from database

emrojo committed:
Load initial state from database and tests

emrojo committed:
Cancel button will close the mapping tool
Added button to be able to reopen it

emrojo committed:
Read only columns

emrojo committed:
Update state from inputs only on blur
Update the value on inputs when the manifest is uploaded
Inputs are uncontrolled react components
Added styling for disabled inputs (scientific name)

emrojo committed:
Collapsable error messages

emrojo committed:
Support for warning messages in server side

emrojo committed:
Add warning validation for unique values in donor id and supplier name

emrojo committed:
Duplication checks now also displays a warning referring the original

emrojo committed:
Readapting tests

emrojo committed:
Update scientific name in the validator

emrojo committed:
Exception handling in proenance state to add some error messages, instead of
using the transformer

emrojo committed:
Corrected warning code

emrojo committed:
Removed eslint

emrojo committed:
Select is not being reset after reloading the file

emrojo committed:
Corrected call to setupInputValue()

emrojo committed:
Corrected call to setupInputValue

emrojo committed:
Reset select value when reapplying a manifest

emrojo committed:
Refactor for message creation into its own module

emrojo committed:
Refactor of ContentAccessor to reduce number of methods in it

emrojo committed:
Reduce the number of arguments for taxon id property validator

emrojo committed:
Refactor message setting for schema validators
Raised code climate duplication mass value for javascript

emrojo committed:
Code climate config settings from manual were already deprecated.

emrojo committed:
Some refactor for code climate

emrojo committed:
Added timeout so capybara does wait enough

pjvv committed:
Not needed file

pjvv committed:
Not needed

emrojo committed:
Updated application.js
